algebra/src/bytes.rs

```diff
@@ -255,16 +287,52 @@ impl FromBytes for bool {
     }
 }
 
+impl<T: ToBytes> ToBytes for Option<T> {
+    #[inline]
+    fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        self.is_some().write(&mut writer)?;
+        if self.is_some() {
+            self.as_ref().unwrap().write(&mut writer)?;
+        }
+        Ok(())
+    }
+}
+
+impl<T: FromBytes> FromBytes for Option<T> {
+    #[inline]
+    fn read<R: Read>(mut reader: R) -> IoResult<Option<T>> {
+        let exists = bool::read(&mut reader)?;
+        Ok(if exists {
+            Some(T::read(&mut reader)?)
+        } else {
+            None
+        })
+    }
+}
+
 impl<T: ToBytes> ToBytes for Vec<T> {
     #[inline]
     fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        (self.len() as u64).write(&mut writer)?;
```

Comment:
What if the Vec is bigger than 2^64 ?

Reply:
Mmm, vector bigger than 2^64 is 16777216 TB of data, do we really need to take care about such situation?

Reply:
So it will always be smaller, then you are using an oversized u64. I would remove the serialization/deserialization for Vec<T> because it highly depends on the use case so we should let the user taking care of it

Reply:
My idea was to have a default implementation which is universal and can fit all needs, and in case we need some optimizations or customization, it can be modified in place. Is it bad? Or other idea: we can always use first 2 bits of saved value as flag what kind of type is used. Something like this:

00 - u8
01 - u16
10 - u32
11 - u64

we will reduce the capacity of each type of course, but will have universal implementation with more efficient memory usage. What do you think?

Same approach can be applied to the usize

Reply:
It may be an idea but would complicate the logic too much (you need to convert from and to bytes/bits when serializing and deserializing). It's difficult for me to see how any of the primitive types may be generic enough (not too big nor too small) to cover a generic implementation like this, I think most likely we are not going to use it in the vast majority of applications, that's why I propose to remove it directly

Reply:
Implementation for general Vec<T> removed